### PR TITLE
Add WMWorkload.validateSiteLisitsUpdate method.

### LINKED ISSUE #12250: Updating priority of a workflow from a script sets the whitelist to an empty list
**Impact of the bug**

pdmv has an automated tool that increases the priority of workflows that are in acquired for longer than N days. This bug caused all these workflows to have an empty whitelist. We will rely on P&R to provide a list of the affected workflows.

We think that workflows in running-closed are not impacted. the ones acquired will likely stay in acquired forever. the impact on workflows in running-open is unclear

**Describe the bug**

pdmv has an automated tool that increases the priority of workflows that are in acquired for longer than N days. This bug caused all these workflows to have an empty whitelist

**How to reproduce it**

Update the priority of a workflow from a script, not the reqmgr web ui.

**Expected behavior**

if we update only the priority from an http call, only priority should be affected

**Additional context and error message**

Alan wrote the following script to replicate the error

<details><summary>setpriority.py</summary>

```python
#!/usr/bin/env python
"""
This script can be used to update the status of a request in ReqMgr2.
"""
from __future__ import print_function, division

import json
import os
import sys
import http.client


def setPrio(url, workflow, newstatus):
    headers = {"Content-type": "application/json",
               "Accept": "application/json"}

    encodedParams = json.dumps({"RequestPriority": newstatus})

    conn = http.client.HTTPSConnection(url, cert_file=os.getenv('X509_USER_PROXY'), key_file=os.getenv('X509_USER_PROXY'))
    conn.request("PUT", "/reqmgr2/data/request/%s" % workflow, encodedParams, headers)
    resp = conn.getresponse()
    if resp.status != 200:
        print("Response status: %s\tResponse reason: %s\tError-detail: %s" % (resp.status, resp.reason, resp.getheader("x-error-detail")))
        print("  FAILED status transition for: %s" % workflow)
    else:
        print("  OK!")
    conn.close()



def main():
#    url = 'cmsweb.cern.ch'
    url = 'cmsweb-testbed.cern.ch'
#    url = 'alancc7-cloud1.cern.ch'
#    url = 'alancc7-cloud2.cern.ch'
#    url = 'cmsweb-k8s-testbed.cern.ch'
#    url = 'cmsweb-test9.cern.ch'
#    url = 'cmsweb-test1.cern.ch'

    args = sys.argv[1:]
    if not len(args) == 2:
        print("usage: python setrequeststatus.py <workflow_name> <priority_value>")
        sys.exit(0)
    wflowName = args[0]
    newprio = int(args[1])
    print("Setting %s priority to  %s" % (wflowName, newprio))
    setPrio(url, wflowName, newprio)


if __name__ == "__main__":
    main()
```

</details>

### PR DESCRIPTION
**Pull request by todor-ivanov**
Fixes #12250

#### Status
ready

#### Description
During the validation of the user provided `request_args` we have a step to mandatory check against the spec arguments  from here: https://github.com/dmwm/WMCore/blob/02834148d19e773a3a99b7e14cb41185218c3116/src/python/WMCore/WMSpec/WMWorkload.py#L2074 

and subsequently the `siteLists` from here: https://github.com/dmwm/WMCore/blob/02834148d19e773a3a99b7e14cb41185218c3116/src/python/WMCore/WMSpec/WMWorkloadTools.py#L294

Unfortunately the so defined function at `WMWorloadTools.validateSiteLists` sets the `sitelists` to empty lists by default (even in the case they are completely missing from the so provided by the user `request_args`): https://github.com/dmwm/WMCore/blob/02834148d19e773a3a99b7e14cb41185218c3116/src/python/WMCore/WMSpec/WMWorkloadTools.py#L247-L259

This causes the following behavior:

During the call to validate a dictionary of arguments provided by the user which is missing the `siteWhitelist` or `SiteBlacklist`  at `ReqMgr2` - here:  https://github.com/dmwm/WMCore/blob/02834148d19e773a3a99b7e14cb41185218c3116/src/python/WMCore/ReqMgr/Utils/Validation.py#L116 

The `siteLists` are set to empty lists before we enter the sequence of steps to update both the request at `ReqMgr` and all the `WorkQueuElements`.

Here is a printout  of this effect produced by an interactive debugging session again together with @mapellidario: 
```
[11/Feb/2025:15:34:51]  Updating request "tivanov_TaskChain_LumiMask_multiRun_SiteListsTest_v7_241210_093449_7256" with these user-provided args: {'RequestPriority': 700}
Before Validation:{'RequestPriority': 700}
After  Validation:{'RequestPriority': 700, 'SiteWhitelist': [], 'SiteBlacklist': []}
Calling _handleNoStatusUpdate
[11/Feb/2025:15:34:51]  Handling request: tivanov_TaskChain_LumiMask_multiRun_SiteListsTest_v7_241210_093449_7256 with CurrentRequest status: running-open
[11/Feb/2025:15:34:51]  reqArgs: {'RequestPriority': 700, 'SiteWhitelist': []}

```

This way the so produced empty lists enter as argument difference in `reqArgsDiff` and are considered as a change to  empty `SiteWhilist` and `SiteBlacklist` and upon that updated in the workflow spec. Causing the so reported bug.

Here is the result after applying the current patch:
```
[11/Feb/2025:15:52:11]  Updating request "tivanov_TaskChain_LumiMask_multiRun_SiteListsTest_v7_241210_093449_7256" with these user-provided args: {'RequestPriority': 900}
Before Validation:{'RequestPriority': 900}
After  Validation:{'RequestPriority': 900}
Calling _handleNoStatusUpdate
[11/Feb/2025:15:52:11]  Handling request: tivanov_TaskChain_LumiMask_multiRun_SiteListsTest_v7_241210_093449_7256 with CurrentRequest status: running-open
[11/Feb/2025:15:52:11]  reqArgs: {'RequestPriority': 900}

```
It is obvious now that the `Request._handleNoStatusUpdate` method is called without the empty siteLists  

#### Is it backward compatible (if not, which system it affects?)
YES

#### Related PRs
The PR introducing the bug was:   https://github.com/dmwm/WMCore/pull/12099 

#### External dependencies / deployment changes
N/A

### CODE REVIEW

File: src/python/Utils/Utilities.py [hunk 1]
```diff
@@ -33,6 +33,8 @@ def makeList(stringList):
     throws a ValueError if the input is not well formed.
     If the stringList is already of type list, then return it untouched.
     """
+    # NOTE: In case of a change to this function beware of side effects which
```
Comment (amaltaro) [inline src/python/Utils/Utilities.py:36]: Given that the `Utils` package is supposed to be agnostic of any WMCore-specific logic, I would suggest to remove this comment, please.

**Review by amaltaro — CHANGES_REQUESTED** [review]
Thank you for providing this fix. In addition to the comment along the code, I also have a comment on the WMWorkloadTools, but I'd rather have it here than along the code. So here it goes.

Even though that looks the way to go, I am not sure whether that is the correct system-wide fix.

When a workflow is created or assigned, the workflow factory is loaded and arguments are:
* set to a default value
* validated the data type
* validated the content
usually following the StdBase + sub-class spec. For the site lists, we use this definition for instance:
https://github.com/dmwm/WMCore/blob/master/src/python/WMCore/WMSpec/StdSpecs/StdBase.py#L1147

so I am afraid that with this change, we will be either causing an error somewhere else or setting these to None as default.

I'd suggest to test/investigate those cases as well.

**Comment by todor-ivanov** [comment]
ok this: 
> When a workflow is created or assigned, the workflow factory is loaded and arguments are:
>
>   * set to a default value
>   * validated the data type
>   * validated the content
 >   usually following the StdBase + sub-class spec. For the site lists, we use this definition for instance:
 >   https://github.com/dmwm/WMCore/blob/master/src/python/WMCore/WMSpec/StdSpecs/StdBase.py#L1147

Tells the whole story. This function has been already developed under the presumption there is no previously set parameters at the workflow. In addition to that, it will not only NOT work for updates of already existing parameters, but it would rather miss eventual `SiteBlacklist` vs. `SiteWhitelists` conflicts between the sets of already existing + new  arguments. Because in the current version of this function it checks for such conflicts only in among the site lists provided with the call but not with the already existing ones. 

I am going to change the code. And will not touch the old one in any way. It would be better in my opinion.

**Review by amaltaro — COMMENTED** [review]
Did you also test a case where the user provides a non-empty list for SiteWhitelist and an empty list for SiteBlacklist?

**Comment by todor-ivanov** [comment]
> Did you also test a case where the user provides a non-empty list for SiteWhitelist and an empty list for SiteBlacklist?

Yes, It behaves exactly as expected. I just did not want to overload the already too long printouts, since this is a case similar to having the `SiteBlacklist` provided while `SiteWhitelist` empty. It will be addressed in the unit tests though.

**Comment by amaltaro** [comment]
@todor-ivanov if this is ready to go, feel free to merge it in and backport it to branch `2.3.8_cmsweb` and follow with a patch release. This should be documented in https://cms-wmcore.docs.cern.ch/wmcore/TaggingAndReleasing/#cutting-a-patch-release
Thanks!
